File: scripts/dynamixel_force_controller_state_publisher.py
#!/usr/bin/env python3
"""
Gripper-force controller (ROS 1, Python 3.8) for a single Dynamixel XM-series servo.
Handles Present/Goal-Current registers as **signed 16-bit two’s-complement**
so all readings and commands use proper units.
"""

import time
from typing import Optional

import numpy as np
import rospy
from diagnostic_msgs.msg import DiagnosticArray, DiagnosticStatus, KeyValue
from geometry_msgs.msg import WrenchStamped
from sensor_msgs.msg import JointState
from std_msgs.msg import Float32
from dynamixel_sdk import *  # Dynamixel SDK

# ──────────────────────────────────────────────────────────────────────────────
# Dynamixel addresses & constants
# ──────────────────────────────────────────────────────────────────────────────
ADDR_PRESENT_POSITION = 132
ADDR_PRESENT_CURRENT  = 126
ADDR_GOAL_CURRENT     = 102
ADDR_TORQUE_ENABLE    = 64
ADDR_GOAL_POSITION    = 116
ADDR_OPERATING_MODE   = 11
ADDR_PRESENT_VELOCITY = 128
ADDR_PRESENT_VOLTAGE  = 144
ADDR_PRESENT_TEMP     = 146

PROTOCOL_VERSION = 2.0

# ──────────────────────────────────────────────────────────────────────────────
# Helper constants / conversion utilities
# ──────────────────────────────────────────────────────────────────────────────
CURRENT_LSB_MA = 2.69   # mA per LSB for Current registers (XM430 datasheet)
DEG_PER_TICK   = 0.088  # ° per position tick (360° / 4096)


def _to_signed(val: Optional[int], bits: int = 16) -> int:
    """Return signed two-complement representation of an unsigned word."""
    if val is None:
        return 0
    if val >= 1 << (bits - 1):          # top bit set ⇒ negative
        val -= 1 << bits
    return val

# ──────────────────────────────────────────────────────────────────────────────
# Application parameters
# ──────────────────────────────────────────────────────────────────────────────
# MIN_POSITION          = 106
# MAX_POSITION          = 190
# NEUTRAL_POSITION      = 120
# FULLY_CLOSED_POSITION = 190

FORCE_THRESHOLD    = 0.5    # N
AUTO_RETURN_SPEED  = 0.1
MIN_RETURN_CURRENT = 150    # mA

TAU      = 64          # angle offset [deg]
K_MOTOR  = 1.769       # Nm/A
B_MOTOR  = 0.2214      # Nm (negative motor offset)
L        = 0.038       # lever arm [m]

# ──────────────────────────────────────────────────────────────────────────────
# ROS node setup
# ──────────────────────────────────────────────────────────────────────────────
rospy.init_node("gripper_force_controller")
rate = rospy.Rate(50)

BAUDRATE_DXL = rospy.get_param("~baudrate_dxl", 57600)
DEVICENAME   = rospy.get_param("~devicename", "/dev/ttyUSB0")
DXL_ID       = rospy.get_param("~dxl_id", 1)

MIN_POSITION = rospy.get_param("~min_ticks")
MAX_POSITION = rospy.get_param("~max_ticks")
NEUTRAL_POSITION = MIN_POSITION + 15   # if you like
FULLY_CLOSED_POSITION = MAX_POSITION    

force_value: Optional[float] = None  # cache of latest force trigger
_last_log: float = 0.0               # throttled log timer

rospy.Subscriber(
    "/gripper_force_trigger",
    Float32,
    lambda msg: globals().__setitem__("force_value", msg.data),
)

joint_pub = rospy.Publisher("/joint_states", JointState, queue_size=10)
force_pub = rospy.Publisher("/gripper_force_stamped", WrenchStamped, queue_size=10)
diag_pub  = rospy.Publisher("/diagnostics", DiagnosticArray, queue_size=10)

# ──────────────────────────────────────────────────────────────────────────────
# Dynamixel port / packet handlers
# ──────────────────────────────────────────────────────────────────────────────
portHandler   = PortHandler(DEVICENAME)
packetHandler = PacketHandler(PROTOCOL_VERSION)

if not portHandler.openPort():
    rospy.logfatal("Failed to open the port %s", DEVICENAME)
    raise SystemExit(1)
if not portHandler.setBaudRate(BAUDRATE_DXL):
    rospy.logfatal("Failed to set baudrate %d", BAUDRATE_DXL)
    raise SystemExit(1)

# Current-based position mode
packetHandler.write1ByteTxRx(portHandler, DXL_ID, ADDR_OPERATING_MODE, 5)
packetHandler.write1ByteTxRx(portHandler, DXL_ID, ADDR_TORQUE_ENABLE, 1)

# ──────────────────────────────────────────────────────────────────────────────
# Utility functions
# ──────────────────────────────────────────────────────────────────────────────

def _throttle_log(msg: str, interval: float = 1.0) -> None:
    global _last_log
    now = time.time()
    if now - _last_log >= interval:
        rospy.loginfo(msg)
        _last_log = now


def read_value(address: int, length: int):
    if length == 4:
        return packetHandler.read4ByteTxRx(portHandler, DXL_ID, address)[0]
    if length == 2:
        return packetHandler.read2ByteTxRx(portHandler, DXL_ID, address)[0]
    if length == 1:
        return packetHandler.read1ByteTxRx(portHandler, DXL_ID, address)[0]
    return None


def read_position_deg() -> Optional[float]:
    raw = read_value(ADDR_PRESENT_POSITION, 4)
    if raw is None:
        rospy.logwarn("Failed to read position")
        return None
    deg = raw * DEG_PER_TICK
    _throttle_log(f"Motor Position: {deg:.2f}°")
    return deg


def read_current_mA() -> float:
    raw = read_value(ADDR_PRESENT_CURRENT, 2)
    if raw is None or raw >= 0xFFFE:  # 0xFFFE/0xFFFF = invalid
        return 0.0
    signed = _to_signed(raw)
    mA = signed * CURRENT_LSB_MA
    _throttle_log(f"Motor Current: {mA:.2f} mA ({signed} LSB)")
    return mA

# ──────────────────────────────────────────────────────────────────────────────
# Publishing helpers
# ──────────────────────────────────────────────────────────────────────────────

def publish_joint_and_diag():
    pos      = read_value(ADDR_PRESENT_POSITION, 4)
    vel      = read_value(ADDR_PRESENT_VELOCITY, 4)
    cur      = read_value(ADDR_PRESENT_CURRENT, 2)
    volt     = read_value(ADDR_PRESENT_VOLTAGE, 2)
    temp     = read_value(ADDR_PRESENT_TEMP, 1)
    goal_cur = read_value(ADDR_GOAL_CURRENT, 2)

    if None in (pos, vel, cur, volt, temp, goal_cur):
        rospy.logwarn("Incomplete telemetry")
        return

    deg        = pos * DEG_PER_TICK
    current_mA = _to_signed(cur) * CURRENT_LSB_MA
    goal_mA    = _to_signed(goal_cur) * CURRENT_LSB_MA

    js = JointState()
    js.header.stamp = rospy.Time.now()
    js.name         = ["gripper"]
    js.position     = [np.deg2rad(deg)]
    js.velocity     = [0.0]
    js.effort       = [current_mA / 1000.0]  # Amps
    joint_pub.publish(js)

    diag               = DiagnosticArray()
    diag.header.stamp  = rospy.Time.now()
    status             = DiagnosticStatus()
    status.name        = "Gripper Motor"
    status.level       = 0
    status.message     = "Nominal"
    status.hardware_id = "XM430-W350"
    status.values      = [
        KeyValue("Position (deg)", f"{deg:.2f}"),
        KeyValue("Velocity", str(vel)),
        KeyValue("Current (mA)", f"{current_mA:.1f}"),
        KeyValue("Voltage (V)", f"{volt * 0.1:.2f}"),
        KeyValue("Temperature (°C)", str(temp)),
        KeyValue("Goal Current (mA)", f"{goal_mA:.1f}"),
    ]
    diag.status.append(status)
    diag_pub.publish(diag)

# ──────────────────────────────────────────────────────────────────────────────
# Motion helpers
# ──────────────────────────────────────────────────────────────────────────────

def move_motor(current_mA: float, position_deg: float):
    pos_val = int(position_deg / DEG_PER_TICK)
    cur_val = int(current_mA / CURRENT_LSB_MA) & 0xFFFF  # two’s-complement encode

    packetHandler.write4ByteTxRx(portHandler, DXL_ID, ADDR_GOAL_POSITION, pos_val)
    packetHandler.write2ByteTxRx(portHandler, DXL_ID, ADDR_GOAL_CURRENT, cur_val)
    _throttle_log(
        f"Commanded {current_mA:.1f} mA → pos {position_deg:.1f}°  (raw {cur_val})"
    )


def move_to_neutral():
    if read_position_deg() is not None:
        move_motor(MIN_RETURN_CURRENT, NEUTRAL_POSITION)


def force_to_current_mA(force_N: float) -> float:
    position = read_position_deg()
    if position is None:
        return 0.0
    angle = position - TAU
    sin_term = np.sin(np.deg2rad(angle)) or 1e-6  # protect against 0
    Nm = force_N * L / sin_term + B_MOTOR
    return (Nm / K_MOTOR) * 1000  # → mA


def grip_object():
    current_mA = force_to_current_mA(force_value)
    move_motor(current_mA, FULLY_CLOSED_POSITION)

# ──────────────────────────────────────────────────────────────────────────────
# Main loop
# ──────────────────────────────────────────────────────────────────────────────
try:
    while not rospy.is_shutdown():
        if force_value is not None:
            # publish wrench
            wrench_msg = WrenchStamped()
            wrench_msg.header.stamp = rospy.Time.now()
            wrench_msg.header.frame_id = "gripper"
            wrench_msg.wrench.force.x = force_value
            force_pub.publish(wrench_msg)

            _throttle_log(f"Force Value: {force_value:.2f} N")
            publish_joint_and_diag()

            pos_deg = read_position_deg()
            if pos_deg is None:
                rate.sleep()
                continue

            within_bounds = MIN_POSITION <= pos_deg <= MAX_POSITION
            if within_bounds and force_value > FORCE_THRESHOLD:
                grip_object()
            else:
                move_to_neutral()

        rate.sleep()

finally:
    packetHandler.write1ByteTxRx(portHandler, DXL_ID, ADDR_TORQUE_ENABLE, 0)
    portHandler.closePort()

File: src/dynamixel_force_controller_state_publisher.cpp
#include <ros/ros.h>
#include <std_msgs/Float32.h>
#include <sensor_msgs/JointState.h>
#include <diagnostic_msgs/DiagnosticArray.h>
#include <diagnostic_msgs/DiagnosticStatus.h>
#include <diagnostic_msgs/KeyValue.h>
#include "dynamixel_sdk/dynamixel_sdk.h"
#include <cmath>

// Dynamixel Addresses and Constants
#define ADDR_PRESENT_POSITION 132
#define ADDR_PRESENT_CURRENT 126
#define ADDR_GOAL_CURRENT 102
#define ADDR_TORQUE_ENABLE 64
#define ADDR_GOAL_POSITION 116
#define ADDR_OPERATING_MODE 11
#define PROTOCOL_VERSION 2.0

#define MIN_POSITION 106.0
#define MAX_POSITION 190.0
#define NEUTRAL_POSITION 120.0
#define FULLY_CLOSED_POSITION 190.0
#define FORCE_THRESHOLD 0.5
#define K_MOTOR 1.769
#define B_MOTOR 0.2214
#define LEVER_ARM_LENGTH 0.038
#define TAU 64.0
#define AUTO_RETURN_CURRENT 150.0

// ROS and Dynamixel variables
float force_value = 0.0;
ros::Time last_log_time;

dynamixel::PortHandler *portHandler;
dynamixel::PacketHandler *packetHandler;

// Utility functions
inline float ticksToDegrees(int32_t ticks) {
    return ticks * 0.088f;
}

inline int32_t degreesToTicks(float degrees) {
    return static_cast<int32_t>(degrees / 0.088f);
}

inline int16_t currentToRaw(float current_mA) {
    return static_cast<int16_t>(current_mA / 2.69f);
}

inline float rawToCurrent(int16_t raw_current) {
    return raw_current * 2.69f;
}

void logThrottled(const std::string &msg) {
    if ((ros::Time::now() - last_log_time).toSec() > 1.0) {
        ROS_INFO_STREAM(msg);
        last_log_time = ros::Time::now();
    }
}

void forceCallback(const std_msgs::Float32::ConstPtr &msg) {
    force_value = msg->data;
}

float readPosition() {
    uint32_t pos_raw;
    packetHandler->read4ByteTxRx(portHandler, 1, ADDR_PRESENT_POSITION, &pos_raw);
    return ticksToDegrees(pos_raw);
}

float readCurrent() {
    uint16_t raw;
    packetHandler->read2ByteTxRx(portHandler, 1, ADDR_PRESENT_CURRENT, &raw);
    int16_t current_signed = raw > 32767 ? raw - 65536 : raw;
    return rawToCurrent(current_signed);
}

void moveMotor(float current_mA, float position_deg) {
    packetHandler->write4ByteTxRx(portHandler, 1, ADDR_GOAL_POSITION, degreesToTicks(position_deg));
    packetHandler->write2ByteTxRx(portHandler, 1, ADDR_GOAL_CURRENT, currentToRaw(current_mA));
}

float forceToCurrentMapping(float force, float gripper_angle_deg) {
    float angle_rad = (gripper_angle_deg - TAU) * M_PI / 180.0f;
    float torque = force * LEVER_ARM_LENGTH / sin(angle_rad);
    float current_A = (torque + B_MOTOR) / K_MOTOR;
    return current_A * 1000.0f;  // mA
}

int main(int argc, char **argv) {
    ros::init(argc, argv, "gripper_force_controller");
    ros::NodeHandle nh("~");

    std::string device_name;
    int baudrate;
    nh.param<std::string>("devicename", device_name, "/dev/ttyUSB0");
    nh.param("baudrate_dxl", baudrate, 57600);

    portHandler = dynamixel::PortHandler::getPortHandler(device_name.c_str());
    packetHandler = dynamixel::PacketHandler::getPacketHandler(PROTOCOL_VERSION);

    if (!portHandler->openPort()) {
        ROS_FATAL("Failed to open the port!");
        return 1;
    }

    if (!portHandler->setBaudRate(baudrate)) {
        ROS_FATAL("Failed to set baudrate!");
        return 1;
    }

    packetHandler->write1ByteTxRx(portHandler, 1, ADDR_OPERATING_MODE, 5);
    packetHandler->write1ByteTxRx(portHandler, 1, ADDR_TORQUE_ENABLE, 1);

    ros::Subscriber force_sub = nh.subscribe("/gripper_force_trigger", 1, forceCallback);
    ros::Publisher joint_pub = nh.advertise<sensor_msgs::JointState>("/joint_states", 1);
    ros::Publisher diag_pub = nh.advertise<diagnostic_msgs::DiagnosticArray>("/diagnostics", 1);
    ros::Rate rate(50);

    while (ros::ok()) {
        float position = readPosition();
        float current = readCurrent();

        if (force_value > FORCE_THRESHOLD && position >= MIN_POSITION && position <= MAX_POSITION) {
            float target_current = forceToCurrentMapping(force_value, position);
            moveMotor(target_current, FULLY_CLOSED_POSITION);
            logThrottled("Gripping object...");
        } else {
            moveMotor(AUTO_RETURN_CURRENT, NEUTRAL_POSITION);
            logThrottled("Returning to neutral...");
        }

        // Publish Joint State
        sensor_msgs::JointState js;
        js.header.stamp = ros::Time::now();
        js.name = {"gripper"};
        js.position = {position * M_PI / 180.0};
        js.effort = {current / 1000.0};
        joint_pub.publish(js);

        // Diagnostics (basic example)
        diagnostic_msgs::DiagnosticArray diag_array;
        diagnostic_msgs::DiagnosticStatus status;
        status.name = "Gripper Motor";
        status.level = diagnostic_msgs::DiagnosticStatus::OK;
        status.message = "Operational";
        status.values.resize(2);
        status.values[0].key = "Position (deg)";
        status.values[0].value = std::to_string(position);
        status.values[1].key = "Current (mA)";
        status.values[1].value = std::to_string(current);
        diag_array.status.push_back(status);
        diag_pub.publish(diag_array);

        ros::spinOnce();
        rate.sleep();
    }

    packetHandler->write1ByteTxRx(portHandler, 1, ADDR_TORQUE_ENABLE, 0);
    portHandler->closePort();

    return 0;
}

